src/apps/bm_devkit/sdi12_example/user_code/user_code.cpp

```diff
@@ -0,0 +1,196 @@
+#include "user_code.h"
+#include "bm_network.h"
+#include "bm_printf.h"
+#include "bm_pubsub.h"
+#include "bristlefin.h"
+#include "bsp.h"
+#include "debug.h"
+#include "lwip/inet.h"
+#include "payload_uart.h"
+#include "sensors.h"
+#include "stm32_rtc.h"
+#include "task_priorities.h"
+#include "uptime.h"
+#include "usart.h"
+#include "util.h"
+#include "exo3s_sensor.h"
+
+#define LED_ON_TIME_MS 20
+#define LED_PERIOD_MS 1000
+#define BYTES_CLUSTER_MS 50
+
+#define DEFAULT_SAMPLE_COUNT 3
+#define MAX_SAMPLE_COUNT 7
+#define SAMPLE_SIZE sizeof(EXO3sample)
+#define MAX_TX_BUFFER_SIZE (uint16_t)(300)
+
+// app_main passes a handle to the user config partition in NVM.
+extern cfg::Configuration *userConfigurationPartition;
+
+uint32_t sample_count = DEFAULT_SAMPLE_COUNT;
+EXO3sample samples[MAX_SAMPLE_COUNT] = {};
+uint32_t current_sample_index = 0;
+uint8_t tx_buffer[MAX_TX_BUFFER_SIZE];  // Buffer for data transmission
+static SondeEXO3sSensor sondeEXO3sSensor;
+
+// A timer variable we can set to trigger a pulse on LED2 when we get payload serial data
+static int32_t ledLinePulse = -1;
+
+
+void transmit_samples(){
+  memset(tx_buffer, 0, sizeof(tx_buffer)); //  Clear the telemetry data buffer.
+  size_t tx_len = SAMPLE_SIZE * sample_count;
+  memcpy(tx_buffer, &samples, tx_len);
+
+  // Get the RTC if available
+  RTCTimeAndDate_t time_and_date = {};
+  rtcGet(&time_and_date);
+  char rtcTimeBuffer[32];
+  rtcPrint(rtcTimeBuffer, &time_and_date);
+
+  // Print hex string of the buffer to the console for debugging.
+  static char tx_hex[MAX_TX_BUFFER_SIZE * 2 + 1];
+  size_t i;
+  for (i = 0; i < tx_len; ++i) {
+    sprintf(&tx_hex[i * 2], "%02X", tx_buffer[i]);
+  }
+  tx_hex[i * 2] = '\0'; // Null-terminate the string
+
+  // print to console
+  printf("[exo3-sonde-sdi12] buffer to send | tick: %llu, rtc: %s, buff: %s\n", uptimeGetMs(), rtcTimeBuffer, tx_hex);
+  // print to spotter console
+  bm_printf(0, "[exo3-sonde-sdi12] buffer to send | tick: %llu, rtc: %s, buff: 0x%s", uptimeGetMs(), rtcTimeBuffer, tx_hex);
+
+  // Transmit over Spotter celluar or Iridium SBD fallback.
+  if (spotter_tx_data(tx_buffer, tx_len, BM_NETWORK_TYPE_CELLULAR_IRI_FALLBACK)) {
+    printf("%llut - %s | Sucessfully sent Spotter transmit data request\n", uptimeGetMs(), rtcTimeBuffer);
+  } else {
+    printf("%llut - %s | Failed to send Spotter transmit data request\n", uptimeGetMs(), rtcTimeBuffer);
+  }
+
+  // resetting the buffer
+  memset(samples, 0, sizeof(samples));
+}
+
+void setup(void) {
+  // samples per send
+  userConfigurationPartition->getConfig("sampleCount", strlen("sampleCount"), sample_count);
```

Comment:
Should we make sure that we limit `sample_count` here to `MAX_TX_BUFFER_SIZE`?
Something on line 78 like:
```
if (sample_count * SAMPLE_SIZE > MAX_TX_BUFFER_SIZE) {
  printf("sampleCount value is larger than max buffer size: %d, limiting sampleCount to: %d", MAX_TX_BUFFER_SIZE, MAX_TX_BUFFER_SIZE / SAMPLE_SIZE);
  sample_count = MAX_TX_BUFFER_SIZE / SAMPLE_SIZE;
}
```

Reply:
I believe the logic was correct before? Moving it below the if statement will not actually limit the sample count

Reply:
Well we are going to read/assign the `sample_count` in from the user configuration partition.
If the user configuration partition sets this value larger than `MAX_TX_BUFFER_SIZE / SAMPLE_SIZE` then we run into the risk of a buffer overflow.

Reply:
Gotcha! I reverted it back. Thanks for explaining.